src/validation_layer.cpp

```diff
@@ -3,7 +3,28 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <validation_layer.h>
-#include <validation.h>
+
+/**
+ * Process an incoming block. This only returns after the best known valid
+ * block is made active. Note that it does not, however, guarantee that the
+ * specific block passed to it has been checked for validity!
+ *
+ * If you want to *possibly* get feedback on whether pblock is valid, you must
+ * install a CValidationInterface (see validationinterface.h) - this will have
+ * its BlockChecked method called whenever *any* block completes validation.
+ *
+ * Note that we guarantee that either the proof-of-work is valid on pblock, or
+ * (and possibly also) BlockChecked will have been called.
+ *
+ * May not be called in a
+ * validationinterface callback.
+ *
+ * @param[in]   pblock  The block we want to process.
+ * @param[in]   fForceProcessing Process this block even if unrequested; used for non-network block sources and whitelisted peers.
+ * @param[out]  fNewBlock A boolean which is set to indicate if the block was first received via this call
+ * @return True if state.IsValid()
+ */
+bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<const CBlock> pblock, bool fForceProcessing, bool* fNewBlock);
```

Comment:
In commit "Limit available scope of ProcessNewBlock to ValidationLayer (move-only)" (377873614b3fa143147ecaba5b750f8be1bdd8a4)

It's unusual that ProcessNewBlock is documented and declared in `validation_layer.cpp` but defined in `src/validation.cpp`. It seems like it will make the documentation hard to find. There are a bunch of other options that seem like they would be better to me:

* Moving `ProcessNewBlock` declaration to `validate_layer.cpp` but moving the documentation to `validate.cpp` near the definition.
* Leaving `ProcessNewBlock` where it is but renaming it to `ProcessNewBlock` internal.
* Making `ProcessNewBlock` a private method of a class that is friends with `ValidationLayer`

Reply:
In commit "Limit available scope of ProcessNewBlock to ValidationLayer (move-only)" (377873614b3fa143147ecaba5b750f8be1bdd8a4)

LOCKS_EXCLUDED(cs_main) annotation seems to have been dropped here.